mlbench_core/lr_scheduler/tensorflow/lr.py

```diff
@@ -0,0 +1,67 @@
+r"""Learning rate scheduling in tensorflow.
+
+The manual_stepping function is taken from :
+
+https://github.com/tensorflow/models/blob/master/research/object_detection/utils/learning_schedules.py
+"""
+
+import tensorflow as tf
+
+
+def manual_stepping(global_step, boundaries, rates, warmup=False):
+    """Manually stepped learning rate schedule.
+
+    This function provides fine grained control over learning rates.  One must
+    specify a sequence of learning rates as well as a set of integer steps
+    at which the current learning rate must transition to the next.  For example,
+    if boundaries = [5, 10] and rates = [.1, .01, .001], then the learning
+    rate returned by this function is .1 for global_step=0,...,4, .01 for
+    global_step=5...9, and .001 for global_step=10 and onward.
+
+    Args:
+        global_step: int64 (scalar) tensor representing global step.
+        boundaries: a list of global steps at which to switch learning
+        rates.  This list is assumed to consist of increasing positive integers.
+        rates: a list of (float) learning rates corresponding to intervals between
+        the boundaries.  The length of this list must be exactly
+        len(boundaries) + 1.
+        warmup: Whether to linearly interpolate learning rate for steps in
+        [0, boundaries[0]].
+
+    Returns:
+        a (scalar) float tensor representing learning rate
+
+    Raises:
+        ValueError: if one of the following checks fails:
+        1. boundaries is a strictly increasing list of positive integers
+        2. len(rates) == len(boundaries) + 1
+        3. boundaries[0] != 0
+    """
+    if any([b < 0 for b in boundaries]) or any(
+            [not isinstance(b, int) for b in boundaries]):
```

Comment:
could be replaced with `not all(isinstance(..)....)` for improved readability(no double-negative)